src/SixLabors.Fonts/GlyphSubstitutionCollection.cs

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using SixLabors.Fonts.Tables.AdvancedTypographic;
+using SixLabors.Fonts.Unicode;
+
+namespace SixLabors.Fonts
+{
+    /// <summary>
+    /// Represents a collection of glyph indices that are mapped to input codepoints.
+    /// </summary>
+    internal sealed class GlyphSubstitutionCollection : IGlyphShapingCollection
+    {
+        /// <summary>
+        /// Contains a map between the index of a map within the collection and its offset.
+        /// </summary>
+        private readonly List<int> offsets = new();
+
+        /// <summary>
+        /// Contains a map between non-sequential codepoint offsets and their glyph ids.
+        /// </summary>
+        private readonly Dictionary<int, GlyphShapingData> glyphs = new();
+
+        /// <summary>
+        /// Gets the number of glyphs ids contained in the collection.
+        /// This may be more or less than original input codepoint count (due to substitution process).
+        /// </summary>
+        public int Count => this.offsets.Count;
+
+        /// <summary>
+        /// Gets or sets the running id of any ligature glyphs contained withing this collection are a member of.
+        /// </summary>
+        public int LigatureId { get; set; } = 1;
+
+        /// <inheritdoc />
+        public ReadOnlySpan<ushort> this[int index] => this.glyphs[this.offsets[index]].GlyphIds;
```

Comment:
So....

A Glyph Id is an identifier for a glyph in a font. Previously Glyph Index and Glyph Id were used interchangeably but I believe that to be incorrect. Lookups for glyphs are performed against a given codepoint and there's no rule that states they should be stored sequentially (_Only that zero is always the fallback glyph_). 

There's more than one at a given index within the collection because the collection is actually more representative of a jagged array (_though we don't use that for API ease of use_). You can have more than one glyph representing a single codepoint following substitution (_e.g. emojis_)